corehq/apps/app_manager/id_strings.py

```diff
@@ -171,9 +171,21 @@ def detail(module, detail_type):
     return u"m{module.id}_{detail_type}".format(module=module, detail_type=detail_type)
 
 
-def menu(module):
+def menu_id(module):
     put_in_root = getattr(module, 'put_in_root', False)
-    return ROOT if put_in_root else u"m{module.id}".format(module=module)
+    if put_in_root:
+        # handle circular calls, if bad module workflow setup
+        return menu_id(module.root_module) if getattr(module, 'root_module', False) else ROOT
+    else:
+        return u"m{module.id}".format(module=module)
+
+
+def menu_root(module):
```

Comment:
The `root` attribute should be empty unless module has a parent AND `module.put_in_root == False`